pkg/cstor/volume/targetserver/server.go

```diff
@@ -0,0 +1,173 @@
+	"os"
+	"strings"
+	"syscall"
+
+	"k8s.io/klog"
+
+	cstorv1alpha1 "github.com/openebs/maya/pkg/cstor/volume/v1alpha1"
+	env "github.com/openebs/maya/pkg/env/v1alpha1"
+	"github.com/pkg/errors"
+)
+
+var (
+	endOfData          = "\r\n"
+	respOk             = "Ok"
+	respErr            = "Err"
+	volumeMgmtUnixSock = "/var/run/volume_mgmt_sock"
+)
+
+// IsSafeToRetry will retrun true if error type is EINTR, EAGAIN or EWOULDBLOCK
+func IsSafeToRetry(err error) bool {
+	// For more information https://golang.org/pkg/syscall/#Errno
+	if err == syscall.EINTR ||
+		err == syscall.EAGAIN ||
+		err == syscall.EWOULDBLOCK {
+		return true
+	}
+	return false
+}
+
+/* Client will send below data to process
+ * Ex JSON data:
+ * {"replicaId":"6061","replicaZvolGuid":"6061",
+ * "volumeName":"vol1","replicationFactor":3,"consistencyFactor":2}
+ * and server will process it and returns Ok in case of not having error else it
+ * returns Err
+ * Response will be either "Ok" or "Err"
+ */
+
+// Reader reads the data from wire untill error or endOfData occurs
+// Reader will break only when client is sending \r\n or EOF occured
+func Reader(r io.Reader) (string, error) {
+	buf := make([]byte, 1024)
+	//collect bytes into fulllines buffer till the end of line character is reached
+	completeBytes := []byte{}
+	for {
+		n, err := r.Read(buf[:])
+		if err != nil {
+			if err == io.EOF {
+				klog.Info("Reached End Of file")
+				break
+			}
+			return "", errors.Wrapf(err, "failed to read data on wire")
+		}
+		if n > 0 {
+			completeBytes = append(completeBytes, buf[0:n]...)
+			if strings.HasSuffix(string(completeBytes), endOfData) {
+				// Since we are ending only when \r\n bytes are present in
+				// completeBytes to extract \r\n need to perform below steps
+				lines := strings.Split(string(completeBytes), endOfData)
+				return lines[0], nil
+			}
+		}
+	}
+	// Code will reach here only when EOF happens
+	return "", errors.Errorf("failed to read data connection closed")
+}
+
+//ServeRequest process the request from the client
+func ServeRequest(conn net.Conn, kubeClient *cstorv1alpha1.Kubeclient) {
+	var err error
+	var readData string
+	defer func() {
+		if err != nil {
+			_, er := conn.Write([]byte(respErr + endOfData))
+			if er != nil {
+				klog.Errorf("failed to inform to client")
+			}
+		} else {
+			_, er := conn.Write([]byte(respOk + endOfData))
+			if er != nil {
+				klog.Errorf("failed to inform to client")
+			}
+		}
+		conn.Close()
+	}()
+	//example readData:
+	// {"replicaId":"6061","replicaZvolGuid":"6061","volumeName":"vol1",
+	// "replicationFactor":3,"consistencyFactor":2}
+	readData, err = Reader(conn)
+	if err != nil {
+		klog.Errorf("failed to read data: {%v}", err)
+		return
+	}
+	replicationData := &cstorv1alpha1.CVReplicationDetails{}
+	err = json.Unmarshal([]byte(readData), replicationData)
+	if err != nil {
+		klog.Errorf("failed to unmarshal replication data {%v}", err)
+		return
+	}
+	csc := &cstorv1alpha1.CStorVolumeConfig{
```

Comment:
lets try to avoid creating one more struct just to pass kubeclient

Reply:
I will take it as a good practice. I will pass as a functional arguments